### PR TITLE
Support for apply_save_mps

### PR DESCRIPTION
**Pull request by merav-aharoni**
### Summary
Add support for the apply_save_mps instruction. This is part of the support for #1062.

### Details and comments
@chriseclectic - I believe we will need a more complicated structure for the mps than what we initially defined. If I understand correctly, we need the `MPSContainer` structure to contain all the information describing the MPS. So what I had in mind is something like the following:
```c++
class MPSContainer {
public:
  MPSContainer() {}

private:
  size_t num_qubits_=0;
  // lambda_size[i] is the length of lambda_struct[i];
  std::vector<size_t> lambda_size_;  
  // lambda_reg_[i] contains the Schmidt coefficients for qubit[i+1]
  std::vector<std::vector<double>> lambda_reg_;
  // q_reg_size_[i] contains the num_rows and num_columns for qreg[i] 
  std::vector<std::pair<size_t, size_t>> q_reg_size_;
  // q_reg_[i] contains 2 complex matrices that make the tensor for qubit i
  std::vector<std::pair<cmatrix_t, cmatrix_t>> q_reg_;
  // The internal ordering of the qubits
  std::vector<uint_t> order_;

};
```
I'd appreciate your input before I continue developing.

### CODE REVIEW

File: qiskit/providers/aer/backends/qasm_simulator.py [hunk 1]
```diff
@@ -297,13 +297,15 @@ class QasmSimulator(AerBackend):
             'kraus', 'roerror', 'snapshot', 'save_expval', 'save_expval_var',
             'save_probabilities', 'save_probabilities_dict',
             'save_density_matrix', 'save_statevector',
-            'save_amplitudes', 'save_amplitudes_sq', 'save_stabilizer'
+            'save_amplitudes', 'save_amplitudes_sq', 'save_stabilizer',
+            'save_matrix_product_state', 'save_mps'
```
Comment (chriseclectic) [inline qiskit/providers/aer/backends/qasm_simulator.py:301]:
```suggestion
            'save_matrix_product_state'
```

File: qiskit/providers/aer/library/save_matrix_product_state.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            simulation.
+        """
+        super().__init__('save_matrix_product_state',
+                         num_qubits,
+                         label,
+                         pershot=pershot,
+                         conditional=conditional)
+
+
+class SaveMatrixProductStateDict(SaveSingleData):
```
Comment (chriseclectic) [inline qiskit/providers/aer/library/save_matrix_product_state.py:49]: I dont think we need this. The MPS is already a custom structure

File: qiskit/providers/aer/library/save_matrix_product_state.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    """
+    qubits = default_qubits(self)
+    instr = SaveMatrixProductState(len(qubits),
+                                   label=label,
+                                   pershot=pershot,
+                                   conditional=conditional)
+    return self.append(instr, qubits)
+
+
+def save_matrix_product_state_dict(self, label="matrix_product_state_dict",
```
Comment (chriseclectic) [inline qiskit/providers/aer/library/save_matrix_product_state.py:103]: Remove

File: qiskit/providers/aer/library/save_matrix_product_state.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    qubits = default_qubits(self)
+    instr = SaveMatrixProductStateDict(len(qubits),
+                                       label=label,
+                                       pershot=pershot,
+                                       conditional=conditional)
+    return self.append(instr, qubits)
+
+
+QuantumCircuit.save_matrix_product_state = save_matrix_product_state
+QuantumCircuit.save_matrix_product_state_dict = save_matrix_product_state_dict
```
Comment (chriseclectic) [inline qiskit/providers/aer/library/save_matrix_product_state.py:131]: Remove

File: src/framework/operations.hpp [hunk 5]
```diff
@@ -86,6 +86,9 @@ inline std::ostream& operator<<(std::ostream& stream, const OpType& type) {
   case OpType::save_statevec_ket:
     stream << "save_statevector_dict";
     break;
+  case OpType::save_mps:
+    stream << "save_mps";
```
Comment (chriseclectic) [inline src/framework/operations.hpp:90]: Should probably change this back to the long name
```suggestion
    stream << "save_matrix_product_state";
```

File: src/simulators/matrix_product_state/matrix_product_state.hpp [hunk 6]
```diff
@@ -559,6 +568,28 @@ void State::apply_ops(const std::vector<Operations::Op> &ops,
...
+  if (last_op) {
+    BaseState::save_data_pershot(result, op.string_params[0],
+				 qreg_.move_to_mps_container(),
+				 op.save_type);
+  } else {
+    BaseState::save_data_pershot(result, op.string_params[0],
+                               qreg_.copy_to_mps_container(),
+                               op.save_type);
+  }
+  //BaseState::save_data_pershot(result, op.string_params[0],
```
Comment (chriseclectic) [inline src/simulators/matrix_product_state/matrix_product_state.hpp:589]: Remove

**Review by chriseclectic — CHANGES_REQUESTED** [review]

File: qiskit/providers/aer/library/save_matrix_product_state.py (hunk 2 above)
Comment (merav-aharoni) [reply qiskit/providers/aer/library/save_matrix_product_state.py]: removed.

File: qiskit/providers/aer/library/save_matrix_product_state.py (hunk 4 above)
Comment (merav-aharoni) [reply qiskit/providers/aer/library/save_matrix_product_state.py]: removed.

File: src/framework/operations.hpp (hunk 5 above)
Comment (merav-aharoni) [reply src/framework/operations.hpp]: done.

**Comment by merav-aharoni** [comment]
@chriseclectic , I believe this is now ready for review + merge.

**Review by vvilpas — DISMISSED** [review]

**Review by chriseclectic — APPROVED** [review]

**Pull request merged**